src/utilities/graphql/transform.ts

```diff
@@ -86,131 +85,330 @@ function nullIfDocIsEmpty(doc: DocumentNode) {
+        const inUse = getInUse(ancestors);
+        if (inUse) {
+          inUse.variables.add(node.name.value);
+        }
       },
+    },
 
-      FragmentSpread: {
-        enter(node) {
-          // Keep track of referenced fragment spreads. This is used to
-          // determine if top level fragment definitions should be removed.
-          fragmentSpreadsInUse[node.name.value] = true;
-        },
+    FragmentSpread: {
+      enter(node, _key, _parent, _path, ancestors) {
+        if (shouldRemoveField(node.directives)) {
+          firstVisitMadeChanges = true;
+          return null;
+        }
+        const inUse = getInUse(ancestors);
+        if (inUse) {
+          inUse.fragmentSpreads.add(node.name.value);
+        }
+        // We might like to remove this FragmentSpread by returning null here if
+        // the corresponding FragmentDefinition node is also going to be removed
+        // by the logic below, but we can't control the relative order of those
+        // events, so we have to postpone the removal of dangling FragmentSpread
+        // nodes until after the current visit of the document has finished.
       },
+    },
 
-      Directive: {
-        enter(node) {
-          // If a matching directive is found, remove it.
-          if (getDirectiveMatcher(directives)(node)) {
-            return null;
-          }
-        },
+    FragmentDefinition: {
+      enter(node, _key, _parent, path) {
+        originalFragmentDefsByPath.set(JSON.stringify(path), node);
       },
-    }),
-  );
+      leave(node, _key, _parent, path) {
+        const originalNode = originalFragmentDefsByPath.get(JSON.stringify(path));
+        if (node === originalNode) {
+          // If the FragmentNode received by this leave function is identical to
+          // the one received by the corresponding enter function (above), then
+          // the visitor must not have made any changes within this
+          // FragmentDefinition node. This fragment definition may still be
+          // removed if there are no ...spread references to it, but it won't be
+          // removed just because it has only a __typename field.
+          return node;
+        }
 
-  // If we've removed fields with arguments, make sure the associated
-  // variables are also removed from the rest of the document, as long as they
-  // aren't being used elsewhere.
-  if (
-    modifiedDoc &&
-    filterInPlace(variablesToRemove, v => !!v.name && !variablesInUse[v.name]).length
-  ) {
-    modifiedDoc = removeArgumentsFromDocument(variablesToRemove, modifiedDoc);
-  }
+        if (
+          // This logic applies only if the document contains one or more
+          // operations, since removing all fragments from a document containing
+          // only fragments makes the document useless.
+          operationCount > 0 &&
+          node.selectionSet.selections.every(selection => (
+            selection.kind === 'Field' &&
```

Comment:
```suggestion
            selection.kind === Kind.FIELD &&
```

[nit] Prefer comparing against `Kind`